### ISSUE TITLE
Tags/MeaninglessTag: allow @param on Struct.new / Data.define constant assignments

### ISSUE THREAD

**Issue by lekemula** [mensfeld/yard-lint#152]
## Description

`Tags/MeaninglessTag` flags `@param` annotations on `Struct.new` / `Data.define` constant assignments. As of [castwide/solargraph#970](https://github.com/castwide/solargraph/pull/970) and [castwide/solargraph#939](https://github.com/castwide/solargraph/pull/939), these annotations are no longer meaningless — Solargraph uses them to type the synthesized accessors.

This is the same issue I previously raised against rubocop-yard ([ksss/rubocop-yard#36](https://github.com/ksss/rubocop-yard/issues/36)), and reproduces here too.

## Reproduction

```ruby
# @param required_gems [Array<String>]
# @param helper_modules [Array<String>]
GemHelpers = Struct.new(:required_gems, :helper_modules, keyword_init: true)
```

```
$ yard-lint --only Tags/MeaninglessTag
[W] gem_helpers.rb:3
    MeaninglessTag: ...
```

Same for `Data.define`:

```ruby
# @param x [Integer]
# @param y [Integer]
Point = Data.define(:x, :y)
```

## Why the current behavior is wrong for this case

The rule's intent — *"@param/@option only make sense on methods"* — predates Solargraph supporting these tags as the type source for `Struct.new`/`Data.define`-generated accessors. For these specific constant assignments, `@param` now carries real, tool-consumed type information and shouldn't be flagged. `@option` is `@param`-specific in Solargraph's support, so it can stay flagged.

## Proposed fix

Same shape as the merge-ready PR I opened on rubocop-yard: [ksss/rubocop-yard#42](https://github.com/ksss/rubocop-yard/pull/42) — when an object's type is `constant` and its assigned value is a `Struct.new(...)` or `Data.define(...)` call (with or without a block), skip `@param` (still report `@option`).

In yard-lint, the relevant code is `lib/yard/lint/validators/tags/meaningless_tag/validator.rb:18-32`:

```ruby
def in_process_query(object, collector)
  object_type = object.type.to_s
  invalid_types = invalid_object_types
  tags_to_check = checked_tags

  return unless invalid_types.include?(object_type)

  object.docstring.tags.each do |tag|
    next unless tags_to_check.include?(tag.tag_name)

    collector.puts "#{object.file}:#{object.line}: #{object.title}"
    collector.puts "#{object_type}|#{tag.tag_name}"
    break
  end
end
```

For `object_type == 'constant'`, the constant's value/source can be inspected (e.g. `object.value`) to detect a `Struct.new(...)` or `Data.define(...)` RHS and skip `@param` accordingly.

## Question for the maintainer

Would you accept a PR along the lines of [ksss/rubocop-yard#42](https://github.com/ksss/rubocop-yard/pull/42) for yard-lint? Happy to draft it — just want to confirm the approach (skip-by-default for `Struct.new`/`Data.define` constants vs. opt-in config flag) before opening.

## Version

- yard-lint: **1.5.1**
- Ruby: 3.3.11

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by lekemula** [comment]
Thank you @mensfeld 🙏

**Comment by mensfeld** [comment]
@lekemula no, thank you for reporting those. My usage scope may be limited, so I am not always able to cover other users cases unless I get solid reports like yours 🙏

**Comment by mensfeld** [comment]
@lekemula all issues you have created have been addressed (I hope so at least) sucessfully. Would you mind testing against master (or if cannot i can craft an rc release) just to double check pretty plz?

**Comment by lekemula** [comment]
Hi @mensfeld, I checked the new changes from `main` and:

1. ✅ `yard-lint --auto-gen-config` now runs without any issues, at least in our project.
2. ✅ `yard-lint --only Tags/MeaninglessTag` does not flag anymore `@param` on Struct/Data definitions
3. ✅ `yard-lint --only Tags/InvalidTypes` now provides more details of the failure
	3.1. ⚠️ Complex Hash values do not seem to be supported yet - will probably extract the "Bonus Point" as a separate issue with more details
	3.2. ⚠️ `generic` keyword does not seem to be recognised - (working in Solargraph but still a YARD proposal  https://github.com/lsegal/yard/issues/1683)

⚠️ are mostly recent developments in YARD/Solargraph similar to `MeaninglessTag`, and I'll probably create separate issues for them with more details. And I might even give them a shot at them myself.

Really appreciate your prompt resolution of the above issues, and I really look forward to integrating this tool into our project soon 🚀

**Comment by mensfeld** [comment]
@lekemula  thank you for checking! Let me see what I can do with the remaining. If you encounter any issues please open appropriate issue reports and i will make sure to tackle them. I should release the newly added fixes this week.

**Comment by mensfeld** [comment]
@lekemula I was not able to reproduce the complex hash values but I added several integration tests just in case. If you have any exact case where it would fail please share and will fix it.

**Comment by mensfeld** [comment]
Re `generic` - as of now it is not part of the yard standard, hence I do not want to support it. Maybe (separate issue) we could at least allow for option to ignore it (do not report, etc) but this will require a bit more thinking.

**Comment by lekemula** [comment]
> Re `generic` - as of now, it is not part of the yard standard, hence I do not want to support it. Maybe (separate issue) we could at least allow for an option to ignore it (do not report, etc) but this will require a bit more thinking.

Makes sense, having the ignore option would be very appreciated for now.

**Comment by mensfeld** [comment]
Let me check if the current architecture can support this type of "skipping" of things the way you utilize it in solargraph. I could maybe write a generic format for it.

**Comment by mensfeld** [comment]
@lekemula there already is `ExtraTypes` but let me check if this would actually support your solargraph cases fully and if so will add more extensive docs for it because I see its not well documented.

**Comment by mensfeld** [comment]
I added relevant tests and updated readme. You should be able to use `ExtraTypes` to silence `generic` until it becomes officially supported. Then it should work out of the box.

**Comment by lekemula** [comment]
That's great! 🎉
